### PR TITLE
[Java][Datetime] Port BaseDatePeriodParser from C# to Java

### PR DESCRIPTION
**Pull request by JuanAr**
- Based on #997. We'll rebase after it is merged to master.
- Skipped 5 tests not supported in java
- Includes a fix to
    - Add Capture class
    -  Replace in MatchGroup: String[] 'captures' for Capture[] 'captures'

### CODE REVIEW

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java [hunk 1]
```diff
@@ -0,0 +1,1602 @@
...
+        LocalDateTime futureStart = DateUtil.safeCreateFromMinValue(futureYear, month, 1);
+        LocalDateTime futureEnd = inclusiveEndPeriod ? futureStart.plusMonths(1).minusDays(1) : futureStart.plusMonths(1);
+
+
+        LocalDateTime pastStart = DateUtil.safeCreateFromMinValue(pastYear, month, 1);
+        LocalDateTime pastEnd = inclusiveEndPeriod ? pastStart.plusMonths(1).minusDays(1) : pastStart.plusMonths(1);
+
+        if (earlyPrefix) {
+            futureEnd = inclusiveEndPeriod
+                ? DateUtil.safeCreateFromMinValue(futureYear,month,15)
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java]: Same. Change in lint rules.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java [hunk 2]
```diff
@@ -0,0 +1,1602 @@
...
+            ret.setSuccess(true);
+            return ret;
+        }
+        return ret;
+    }
+
+    private DateTimeResolutionResult parseWeekOfDate(String text, LocalDateTime referenceDate) {
+        DateTimeResolutionResult ret = new DateTimeResolutionResult();
+        Optional<Match> match = Arrays.stream(RegExpUtility.getMatches(config.getWeekOfRegex(),text)).findFirst();
+        List<ExtractResult> ex = config.getDateExtractor().extract(text, referenceDate);
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/parsers/BaseDatePeriodParser.java]: Nit: line break format/code style.

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/utilities/DurationParsingUtil.java [hunk 3]
```diff
@@ -1,134 +1,130 @@
...
-            case "M":
-                result = true;
-                break;
-            case "S":
-                result = true;
-                break;
-        }
-
-        return result;
+  public static boolean isTimeDurationUnit(String unitStr) {
```
Comment (tellarin) [inline Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/utilities/DurationParsingUtil.java]: Nit: Should be 4 spaces as others.

**Comment by JuanAr** [comment]
@tellarin, rebased branch with master.

**Pull request merged**